### ISSUE TITLE
Database connection closures causing delays and timeouts in tomcat

### ISSUE THREAD

**Issue by daveram** [prometheus/jmx_exporter#502]
We're running tomcat with a oracle database backend using jmx_exporter as a java agent on our tomcat instances, and over the past couple months have noticed many of our jmx_exporter scrapes from Prometheus are timing out.  We also use glowroot https://github.com/glowroot/glowroot for detailed code traces and there we found that the metrics endpoint from jmx_exporter is getting blocked by ... oracle.jdbc.driver.PhysicalConnection.isClosed(PhysicalConnection.java:2573)

![jmx_exporter_response_times](https://user-images.githubusercontent.com/8410976/83722347-ab6d7480-a5f1-11ea-80e7-97f0eb033425.PNG)
![jmx_exporter_trace](https://user-images.githubusercontent.com/8410976/83722349-ac060b00-a5f1-11ea-954d-add3725fec56.PNG)

We have multiple datasources defined in tomcat, and noticed that jmx_exporter is running the following query "SELECT SYS_CONTEXT ('USERENV', 'CURRENT_SCHEMA')  FROM DUAL" but we don't know is this for all datasources every time jmx_exporter is scraped, only some, or some other query that's internal to jmx_exporter?

This would help us determine where the issue may lie.  Current suspects include network and database as restarting tomcat makes no difference, however I also suspect this may be JMX not handling heavy traffic load and a large amount of stats we're pulling.  Note whitelists/blacklists are in place to ensure we aren't just pulling everything, but it's still a larger list

**Comment by brian-brazil** [comment]
The JMX exporter will try to hit every mBean that is exposed, and given that stack trace it looks like the issue is in your database code.

**Comment by daveram** [comment]
Thanks Brian, we finally found the culprit figured I'd share the results here for anyone else running into a similar problem.

Further stack traces showed we had a long-running query which was running every 30 seconds and taking some time to run.  This was locking up threads while waiting for the query to finish.  I noticed that in our tomcat 9.x mbeans we have two spots for jdbc datasources, Catalina:type=DataSource and tomcat.jdbc:*, within the tomcat.jdbc one I found it's polling each connection for connection closed true/false which means it has to wait for the thread to unlock before it can return those mbeans.

Just added tomcat.jdbc:* to blacklistObjectNames and within seconds we started getting all jmx_exporter transactions back to around 100ms instead of running over the 10s prometheus scrape timeout.

It's worth noting that we don't even use these mbeans in our jmx_exporter rules config, but as it queries every mbean available by default we were still seeing the delay on prometheus scrapes.  Happy to report blacklistObjectNames works as intended for this and we're good to go
